source/common/secret/secret_manager_impl.h

```diff
@@ -19,22 +19,42 @@ class SecretManagerImpl : public SecretManager, Logger::Loggable<Logger::Id::ups
 
   void addOrUpdateSecret(const std::string& config_source_hash,
                          const envoy::api::v2::auth::Secret& secret) override;
-  const Ssl::TlsCertificateConfig* findTlsCertificate(const std::string& config_source_hash,
-                                                      const std::string& name) const override;
+  const Ssl::TlsCertificateConfigSharedPtr
+  findTlsCertificate(const std::string& config_source_hash, const std::string& name) const override;
+
   std::string addOrUpdateSdsService(const envoy::api::v2::core::ConfigSource& config_source,
                                     std::string config_name) override;
 
+  void registerTlsCertificateConfigCallbacks(const std::string& config_source_hash,
+                                             const std::string& secret_name,
+                                             SecretCallbacks& callback) override;
+
 private:
   Server::Instance& server_;
-  // map hash code of SDS config source and SdsApi object
+
+  // map hash code of SDS config source and SdsApi object.
   std::unordered_map<std::string, SdsApiPtr> sds_apis_;
   mutable std::shared_timed_mutex sds_api_mutex_;
 
   // Manages pairs of name and Ssl::TlsCertificateConfig grouped by SDS config source hash.
   // If SDS config source hash is empty, it is a static secret.
-  std::unordered_map<std::string, std::unordered_map<std::string, Ssl::TlsCertificateConfigPtr>>
+  std::unordered_map<std::string,
+                     std::unordered_map<std::string, Ssl::TlsCertificateConfigSharedPtr>>
       tls_certificate_secrets_;
   mutable std::shared_timed_mutex tls_certificate_secrets_mutex_;
+
+  // callback functions for secret update
+  // "config source hash": {
+  //   "secret name":
+  //      secret,
+  //      [{callback}]
+  //   ]
+  // }
+  std::unordered_map<std::string,
+                     std::unordered_map<std::string, std::pair<Ssl::TlsCertificateConfigSharedPtr,
```

Comment:
Why do we need to store secret here?  It can be just

map of "hash + name " -> vector<callback>